File: 46-permutations/46-permutations.py
class Solution(object):
    def permute(self, nums):
        self.result = []
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n = len(nums)
        def bfs(i, cur):
            if i == n - 1:
                self.result.append(cur[:])
                return
            for j in range(i, n):
                cur[i], cur[j] = cur[j], cur[i]
                bfs(i+1, cur)
                cur[i], cur[j] = cur[j], cur[i]
        bfs(0,nums)
        return self.result
        

File: 46-permutations/46-permutations.java
class Solution {
    private Set<Integer> visited = new HashSet<>();
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new LinkedList<>();
        
        permuteRes(nums, res, new LinkedList<>());
        return res;
    }
    
    public void permuteRes(int[] nums, List res, LinkedList temp){
        if (temp.size() == nums.length){
            res.add((List) temp.clone());
            return;
        }
        for(int i = 0; i < nums.length; i++){
            if(visited.contains(i))continue;
            visited.add(i);
            temp.addLast(nums[i]);
            permuteRes(nums, res, temp);
            temp.pollLast();
            visited.remove(i);
        }
        
        
        
        
    }
}